contracts/integrations/AutomateReadyUpgradeable.sol

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./Types.sol";
+import {
+    Initializable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+/**
+ * @dev Inherit this contract to allow your upgradeable smart contract to
+ * - Make synchronous fee payments.
+ * - Have call restrictions for functions to be automated.
+ */
+//solhint-disable func-name-mixedcase
+// solhint-disable private-vars-leading-underscore
+abstract contract AutomateReadyUpgradeable is Initializable {
+    IAutomate public immutable automate;
+    address internal constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    address internal feeCollector;
+    address public dedicatedMsgSender;
+
+    /**
+     * @dev
+     * Only tasks created by _taskCreator defined in constructor can call
+     * the functions with this modifier.
+     */
+    modifier onlyDedicatedMsgSender() {
+        require(msg.sender == dedicatedMsgSender, "Only dedicated msg.sender");
+        _;
+    }
+
+    constructor(address _automate) {
+        automate = IAutomate(_automate);
```

Comment:
One question @brandonchuah , if someone has a live upgradable contract, which is not inheriting `AutomateReadyUpgrdable` yet, and want to upgrade it to extend `AutomateReadyUpgradable`.
Then this constructor won't be called?

Reply:
Usually, the upgradable contract templates are used for new undeployed contracts, it is not advisable to inherit `AutomateReadyUpgrdable` for an already deployed contract as there will be potential clashing of storage slots. A more customized approach is recommended. 

But this constructor will still be called by the parent contract when the upgradable contract is redeployed.

Reply:
Ok, thanks for the info @brandonchuah